### PR TITLE
Use semaphores for thread parking on Apple platforms

### PR DESCRIPTION
**Pull request by joboet**
Currently we use a mutex-condvar pair for thread parking on Apple systems. Unfortunately, `pthread_cond_timedwait` uses the real-time clock for measuring time, which causes problems when the system time changes. The parking implementation in this PR uses a semaphore instead, which measures monotonic time by default, avoiding these issues. As a further benefit, this has the potential to improve performance a bit, since `unpark` does not need to wait for a lock to be released.

Since the Mach semaphores are poorly documented (I could not find availability or stability guarantees for instance), this uses a [dispatch semaphore](https://developer.apple.com/documentation/dispatch/dispatch_semaphore?language=objc) instead. While it adds a layer of indirection (it uses Mach semaphores internally), the overhead is probably negligible.

Tested on macOS 12.5.

r? @thomcc

### CODE REVIEW

File: library/std/src/sys/unix/thread_parker/darwin.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+type dispatch_semaphore_t = *mut crate::ffi::c_void;
+type dispatch_time_t = u64;
+
+const DISPATCH_TIME_NOW: dispatch_time_t = 0;
+const DISPATCH_TIME_FOREVER: dispatch_time_t = !0;
+
+#[link(name = "System", kind = "dylib")]
+extern "C" {
+    fn dispatch_time(when: dispatch_time_t, delta: i64) -> dispatch_time_t;
```
Comment (thomcc) [inline library/std/src/sys/unix/thread_parker/darwin.rs:29]: Arguably these bindings should be in libc. I guess a reason not to put it there is that libdispatch can be used on other OSes (linux, windows, freebsd, ...), but needs a separate lib -- it's only part of ~~libc~~ libSystem on darwin-based OSes. Anyway, this is tiny, so I'm not that concerned.

**Review by thomcc — COMMENTED** [review]
I prefer us using dispatch semaphores to mach ones here. Mach semaphores are much heavier weight, and I'm not sure it's as well-integrated into the QOS system as dispatch_semaphore (I think that stuff is done somewhat manually from libdispatch).

The impl looks like a nearly complete match to the futex version (makes sense given that the futex is being used as a semaphore). That makes it pretty easy to review, not that it's unclear anyway.

I'm very happy to see another platform have the SeqCst barrier removed -- not because of performance though, but because we don't guarantee it's SeqCst, so I'd been a bit worried that people would start to rely on `park()` providing that to avoid races! This impl provides correct/optimal orderings that we guarantee, which is nice to see.

I'd like to avoid breaking miri for now, so won't R+ quite yet (and there are a few other nits). I may also run some benchmarks in the meantime (I know `usync` has benchmarks and heavily leverages stdlib parking, so I'm curious if it's impacted -- there are probably others too). A concern is that dispatch_semaphores are... quite heavyweight. For one, they're fair, which (at least if you implement a Mutex using a semaphore and counter), often makes their use come with a bit of a throughput loss...

I hope not though, it's a really nice impl with several benefits!

File: library/std/src/sys/unix/thread_parker/darwin.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+        let timeout = dispatch_semaphore_wait(self.semaphore, timeout) != 0;
+
+        let state = self.state.swap(EMPTY, Acquire);
+        if state == NOTIFIED && timeout {
+            // If the state was NOTIFIED but semaphore_wait returned without
+            // decrementing the count because of a timeout, it means another
+            // thread is about to call semaphore_signal. We must wait for that
+            // to happen to ensure the semaphore count is reset.
+            while dispatch_semaphore_wait(self.semaphore, DISPATCH_TIME_FOREVER) != 0 {}
```
Comment (joboet) [inline library/std/src/sys/unix/thread_parker/darwin.rs:102]: Note that this is technically a priority hole: if a high-priority thread uses `park_timeout` while a low priority thread tries to unpark it, a medium-priority thread could prevent the unpark from completing, so the high-priority thread would wait indefinitely. I doubt that anyone relies on this, but still... By the way, libdispatch itself is [vulnerable](https://github.com/apple/swift-corelibs-libdispatch/blob/469c8ecfa0011f5da23acacf8658b6a60a899a78/src/semaphore.c#L119-L133) to this.

File: library/std/src/sys/unix/thread_parker/mod.rs [hunk 3]
```diff
@@ -11,7 +11,15 @@
 )))]
 
 cfg_if::cfg_if! {
-    if #[cfg(target_os = "netbsd")] {
+    if #[cfg(any(
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply library/std/src/sys/unix/thread_parker/mod.rs]: Thanks! Yes if a fallback exists anyway, using that under Miri for now would be best.

I opened an issue on the Miri side to track this: https://github.com/rust-lang/miri/issues/2589.

File: library/std/src/sys/unix/thread_parker/darwin.rs (hunk 2 above)
Comment (thomcc) [reply library/std/src/sys/unix/thread_parker/darwin.rs:102]: Yeah, that's a fundamental issue with stuff like semaphores.

**Comment by thomcc** [comment]
I'm not able to find any throughput issues in benchmarks of this, so we should be good to go.

@bors r+

**Pull request merged**